File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        const int N = nums.size();
        vector<int> dp(N, -1);
        
        return minJumpReq(nums, 0, N, dp);
    }
    
private:
    int minJumpReq(vector<int>& nums, int curPos, const int N, vector<int>& dp) {
        
        if(curPos == N -1) return 0;
        if(dp[curPos] != -1) return dp[curPos];
        
        int minJump = INT_MAX -1;
        
        for(int i = 1; i <= nums[curPos]; i++) {
            if(curPos +i >= N) break;
            minJump = min(minJump, 1 +minJumpReq(nums, curPos +i, N, dp));
        }
        
        return dp[curPos] = minJump;
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    private static final int BIG_NO = 1000000000;
    
    public int jump(int[] nums) {
        return minJump(nums);
    }
    
    private int minJump(int[] nums) {
        
        final int N = nums.length;
        int[] dp = new int[N];
        
        for(int pos = N -2; pos >= 0; pos--) {
            
            int minCost = BIG_NO;
            for(int k = pos +1; k <= pos +nums[pos]; k++) {
                int cost = (k >= 0 && k < N)? dp[k] : 0;
                cost += 1;
                
                minCost = Math.min(minCost, cost);
            }

            dp[pos] = minCost;
        }
        
        return dp[0];
        
//         if(pos >= nums.length -1) {
//             return 0;
//         }
//         // if(nums[pos] == 0) {
//         //     return BIG_NO;
//         // }
        
//         int minCost = BIG_NO;
//         for(int k = pos +1; k <= pos +nums[pos]; k++) {
//             int cost = 1 +minJump(nums, k);
//             minCost = Math.min(minCost, cost);
            
//         }
        
//         return minCost;
    }
    
}